adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleBanner.m

```diff
@@ -0,0 +1,275 @@
+// limitations under the License.
+
+#import "GADMediationVungleBanner.h"
+#include <stdatomic.h>
+#import "GADMAdapterVungleRouter.h"
+#import "GADMAdapterVungleBiddingRouter.h"
+#import "GADMAdapterVungleUtils.h"
+
+@interface GADMediationVungleBanner () <GADMAdapterVungleDelegate, GADMediationBannerAd>
+@end
+
+@implementation GADMediationVungleBanner {
+  /// Ad configuration for the ad to be loaded.
+  GADMediationBannerAdConfiguration *_adConfiguration;
+
+  /// The completion handler to call when an ad loads successfully or fails.
+  GADMediationBannerLoadCompletionHandler _adLoadCompletionHandler;
+
+  /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
+  id<GADMediationBannerAdEventDelegate> _delegate;
+
+  /// The requested ad size.
+  GADAdSize _bannerSize;
+
+  /// Indicates whether a banner ad is loaded.
+  BOOL _isAdLoaded;
+
+  /// Indicates whether the banner ad finished presenting.
+  BOOL _didBannerFinishPresenting;
+    
+  /// UIView to send to Google's view property and for Vungle to mount the ad
+  UIView *_bannerView;
+}
+
+@synthesize desiredPlacement;
+@synthesize bannerState;
+@synthesize uniquePubRequestID;
+@synthesize isRefreshedForBannerAd;
+@synthesize isRequestingBannerAdForRefresh;
+@synthesize view;
+@synthesize isAdLoaded;
+
+- (void)dealloc {
+    [self cleanUp];
+}
+
+- (nonnull instancetype)initWithAdConfiguration:(nonnull GADMediationBannerAdConfiguration*)adConfiguration
+                              completionHandler:(nonnull GADMediationBannerLoadCompletionHandler)completionHandler {
+  self = [super init];
+  if (self) {
+    _adConfiguration = adConfiguration;
+    _bannerSize = [self vungleAdSizeForAdSize:[adConfiguration adSize]];
+      
+    VungleAdNetworkExtras *networkExtras = adConfiguration.extras;
+    self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:adConfiguration.credentials.settings networkExtras:networkExtras];
+    self.uniquePubRequestID = [networkExtras.UUID copy];
+
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationBannerLoadCompletionHandler origAdLoadHandler = [completionHandler copy];
+    /// Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationBannerAdEventDelegate>(
+      id<GADMediationBannerAd> ad, NSError *error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationBannerAdEventDelegate> delegate = nil;
+      if (origAdLoadHandler) {
+        delegate = origAdLoadHandler(ad, error);
+      }
+      origAdLoadHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestBannerAd {
+  if (!IsGADAdSizeValid(_bannerSize)) {
+    NSString *errorMessage = [NSString stringWithFormat:@"Unsupported ad size requested for Vungle. Size: %@", NSStringFromGADAdSize(_bannerSize)];
```

Comment:
Nit, can you update the error message to:
```
The requested banner size: %@ is not supported by Vungle SDK.
```
?

Reply:
Updated.